enhancements/ingress/optional-ingress.md

```diff
@@ -0,0 +1,274 @@
+  - Make the OpenShift installer tolerate the absence of the default ingress controller.
+  - Make the cluster authentication operator tolerate the absence of the default ingress controller.
+  - Make the cluster monitoring operator tolerate the absence of the default ingress controller.
+- Disable the Route API or try to tolerate the absence of the default ingress controller.
+- Implement the ingress operator disabling in the HyperShift's control plane operator.
+
+## Proposal
+
+### User Stories
+
+#### Disable the ingress on HyperShift
+
+The user of the HyperShift API wants to use an AWS ALB to route the traffic to its example application deployed on a hosted cluster.
+
+The user sets the capabilities in one of the HyperShift resources (**note**: the capability API is yet to be defined in HyperShift):
+```yaml
+apiVersion: hypershift.openshift.io/v1beta1
+kind: HostedCluster
+spec:
+  capabilities:
+    additionalDisabledCapabilities:
+    - ingress
+    baselineCapabilitySet: vCurrent
+...
+```
+
+The HyperShift's control plane operator skips the deployment of the ingress operator and creates a hosted cluster with the `ClusterVersion` as follows:
+```yaml
+apiVersion: config.openshift.io/v1
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  capabilities:
+    additionalEnabledCapabilities:
+    - marketplace
+    - openshift-samples
+    - CSISnapshot
+    - Console
+    - Insights
+    - Storage
+    - baremetal
+    - marketplace
+    - openshift-samples
+    baselineCapabilitySet: None
+...
+```
+Note the missing `ingress` capability in the list of the additional enabled capabilities.
+
+To be able to route the traffic to its example application the user installs [AWS Load Balancer Operator](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/aws-load-balancer-operator.md) and sets up the default `AWSLoadBalancerController`:
+```yaml
+apiVersion: networking.olm.openshift.io/v1
+kind: AWSLoadBalancerController
+metadata:
+  name: cluster
+spec:
+  credentials:
+    name: aws-sts-credentials
+```
+Then the user creates the `Ingress` resource for the default (`alb`) ingress class:
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: example
+  annotations:
+    alb.ingress.kubernetes.io/scheme: internet-facing
+spec:
+  ingressClass: alb
+  rules:
+  - host: example.com
+    http:
+        paths:
+        - path: /
+          backend:
+            service:
+              name: example
+              port:
+                number: 80
+```
+The aws load balancer controller will provision an ALB.
+
+Also, the user can install [External DNS Operator](https://github.com/openshift/enhancements/blob/master/enhancements/dns/externaldns-operator.md) to set up a convenient DNS name for its service (ingress support is coming soon).
+
+### Workflow Description
+TBD
+
+### API Extensions
+N/A - no api extensions are being introduced in this proposal
```

Comment:
This enhancement proposes adding the `capabilities` property to the HostedCluster (and presumably also the HostedControlPlane) `spec` structures, and modifications to existing CRDs are [part of this section's target subject][1].  I'm fine with "Skipping this section while we're scaffolding alternatives" or some other placeholder, and I'm not an approver, so it doesn't matter if I'm fine with something or not, but I'd have expected the `capabilities` additions to get discussed in more detail here.

[1]: https://github.com/openshift/enhancements/blob/a0482b15dfb9338ebbd6e7c6d711302febf78b99/guidelines/enhancement_template.md?plain=1#L187

Reply:
The API for `HostedCluster` was made up just for the demonstration purpose. I'm not proposing to add it, I'm just trying to guess how it may look like. Maybe I should not add `HostedCluster` yaml to this EP at all, to avoid the confusion.
Sorry, I wanted to add you as approver, I just forgot, sorry. Correcting it...

Reply:
Removed the HostedCluster yaml to avoid confusion. Added `wking` as approver.

Reply:
Agree with Trevor, it'd be good to see API details on how the new capability https://github.com/openshift/enhancements/blob/c5fc7f64a8fc43397c5e81d25c867038d2eabbae/enhancements/installer/component-selection.md#implementation-detailsnotesconstraints and how we plan to expose them in Hypershift API.

Reply:
> Agree with Trevor, it'd be good to see API details on how the new capability

The details of how the capability API can be used in the installer or CVO are described in [the component selection EP](https://github.com/openshift/enhancements/blob/master/enhancements/installer/component-selection.md). This enhancement doesn't change it in any way, just adds a new capability. How to add a new capability is [detailed in the same EP](https://github.com/openshift/enhancements/blob/master/enhancements/installer/component-selection.md#how-to-implement-a-new-capability).

> and how we plan to expose them in Hypershift API.

That's a decision which needs to be taken by the engineers responsible for HyperShift API. Probably deserves a dedicated EP done by the HyperShift engineers.

Reply:
I agree that this EP doesn't need to describe the capabilities API itself or how capabilities get specified on HyperShift.  However, it seems to me that the name of the capability itself is an API, for the following reasons:

* Implementing this capability requires a new API definition in openshift/api.
* The ingress operator uses the API definition to specify which manifests to install when the capability is enabled.  
* The cluster-admin may use the API definition to check whether the capability is enabled on a cluster.  

In my opinion, it would be useful to document the new capability name here, under "API Extensions".  (I would argue that *any* EP that defines a new capability should define that new capability under "API Extensions".)

Reply:
The update of the cluster version API is added to `API Extensions` chapter.